tl.go

```diff
@@ -0,0 +1,177 @@
+	DefaultCtxKey   = "tl"
+)
+
+type tlCtxKey string
+
+type TL interface {
+	UploadLogEntry(ctx context.Context, pkToken *pktoken.PKToken, payload, signature []byte, signer crypto.Signer) ([]byte, error)
+	VerifyLogEntry(ctx context.Context, entryBytes []byte) error
+}
+
+type MockTL struct {
+	UploadLogEntryFunc func(ctx context.Context, pkToken *pktoken.PKToken, payload, signature []byte, signer crypto.Signer) ([]byte, error)
+	VerifyLogEntryFunc func(ctx context.Context, entryBytes []byte) error
+}
+
+func (tl *MockTL) UploadLogEntry(ctx context.Context, pkToken *pktoken.PKToken, payload, signature []byte, signer crypto.Signer) ([]byte, error) {
+	if tl.UploadLogEntryFunc != nil {
+		return tl.UploadLogEntryFunc(ctx, pkToken, payload, signature, signer)
+	}
+	return nil, nil
+}
+
+func (tl *MockTL) VerifyLogEntry(ctx context.Context, entryBytes []byte) error {
+	if tl.VerifyLogEntryFunc != nil {
+		return tl.VerifyLogEntryFunc(ctx, entryBytes)
+	}
+	return nil
+}
+
+type RekorTL struct{}
+
+// UploadLogEntry submits a PK token signature to the transparency log
+func (tl *RekorTL) UploadLogEntry(ctx context.Context, pkToken *pktoken.PKToken, payload, signature []byte, signer crypto.Signer) ([]byte, error) {
+	// generate self-signed x509 cert to wrap PK token
+	pubCert, err := CreateX509Cert(pkToken, signer)
+	if err != nil {
+		return nil, fmt.Errorf("Error creating x509 cert: %w", err)
+	}
+
+	// generate hash of payload
+	hasher := sha256.New()
+	hasher.Write(payload)
+
+	// upload entry
+	rekorClient, err := rclient.GetRekorClient(DefaultRekorURL)
+	if err != nil {
+		return nil, fmt.Errorf("Error creating rekor client: %w", err)
+	}
+	entry, err := cosign.TLogUpload(ctx, rekorClient, signature, hasher, pubCert)
+	if err != nil {
+		return nil, fmt.Errorf("Error uploading tlog: %w", err)
+	}
+	entryBytes, err := entry.MarshalBinary()
+	if err != nil {
+		return nil, fmt.Errorf("error marshalling TL entry: %w", err)
+	}
+	return entryBytes, nil
+}
+
+// VerifyLogEntry verifies a transparency log entry
+func (tl *RekorTL) VerifyLogEntry(ctx context.Context, entryBytes []byte) error {
+	entry := new(models.LogEntryAnon)
+	err := entry.UnmarshalBinary(entryBytes)
+	if err != nil {
+		return fmt.Errorf("error failed to unmarshal TL entry: %w", err)
+	}
+	err = entry.Verification.Validate(strfmt.Default)
+	if err != nil {
+		return fmt.Errorf("TL entry failed validation: %w", err)
+	}
+
+	verifier, err := loadVerifier()
+	if err != nil {
+		return fmt.Errorf("error failed to load TL verifier: %w", err)
+	}
+	err = verify.VerifyLogEntry(ctx, entry, verifier)
+	if err != nil {
+		return fmt.Errorf("TL entry failed verification: %w", err)
+	}
+	return nil
+}
+
+func loadVerifier() (signature.Verifier, error) {
+	rekorClient, err := rclient.GetRekorClient(DefaultRekorURL)
+	if err != nil {
+		return nil, fmt.Errorf("error creating rekor client: %w", err)
+	}
+	// fetch key from server
+	keyResp, err := rekorClient.Pubkey.GetPublicKey(nil)
```

Comment:
This should be replaced by using sigstore's TUF for the Rekor public key instead of Rekor's pubkey endpoint.

Reply:
The [rekor-cli doesn't use TUF yet](https://github.com/sigstore/rekor/blob/7788eb3215270c218bcbace74b2f028ad8d21c9c/cmd/rekor-cli/app/verify.go#L164-L169) but the [cosign rekor client does](https://github.com/sigstore/cosign/blob/7763aefff1af577878fdd3befd1af8dcb1befc6d/pkg/cosign/tlog.go#L135-L147)

Reply:
fixed [884a1d6](https://github.com/openpubkey/signed-attestation/pull/4/commits/884a1d6612d26d4c8e4bf3e29a6374f4b5b074d5)